libraries/AP_Camera/AP_RunCam.cpp

```diff
@@ -0,0 +1,831 @@
+    if (param_len > 0) {
+        *b++ = param;
+    }
+
+    uint8_t crc = 0;
+    for (const uint8_t *ptr = _buffer; ptr < b; ++ptr) {
+        crc = crc8_dvb_s2(crc, *ptr);
+    }
+
+    *b = crc;
+
+    // send data if possible
+    uart->write(_buffer, 3 + param_len);
+}
+
+// crc functions
+uint8_t AP_RunCam::crc8_dvb_s2(uint8_t crc, uint8_t a)
+{
+    crc ^= a;
+    for (uint8_t i = 0; i < 8; ++i) {
+        if (crc & 0x80) {
+            crc = (crc << 1) ^ 0xD5;
+        } else {
+            crc = crc << 1;
+        }
+    }
+    return crc;
+}
+
+uint8_t AP_RunCam::crc8_high_first(uint8_t *ptr, uint8_t len)
+{
+    uint8_t crc = 0x00;
+    while (len--) {
+        crc ^= *ptr++;
+        for (uint8_t i = 8; i > 0; --i) {
+            if (crc & 0x80)
+                crc = (crc << 1) ^ 0x31;
+            else
+                crc = (crc << 1);
+        }
+    }
+    return (crc);
+}
+
+// handle a device info response
+void AP_RunCam::parse_device_info(RunCamResponse* response)
+{
+    _protocol_version = (RCProtocolVersion)response->_recv_buf[1];
+
+    uint8_t featureLowBits = response->_recv_buf[2];
+    uint8_t featureHighBits = response->_recv_buf[3];
+    _features = (featureHighBits << 8) | featureLowBits;
+    _is_ready = true;
+    gcs().send_text(MAV_SEVERITY_INFO, "RunCam device initialized, features 0x%04X\n", _features.get());
+}
+
+// wait for the RunCam device to be fully ready
+bool AP_RunCam::is_initialized()
+{
+    if (_is_initialized) {
+        return true;
+    }
+
+    if (!_is_ready) {
+        return false;
+    }
+
+    if (AP_HAL::millis() > _boot_delay_ms) {
+        _is_initialized = true;
+
+        // make sure the recording state reflects the control button
+        if (_video_recording) {
+            simulate_camera_button(RCDEVICE_PROTOCOL_CAM_CTRL_START_RECORDING);
+        }
+        else {
+            simulate_camera_button(RCDEVICE_PROTOCOL_CAM_CTRL_STOP_RECORDING);
+        }
+        debug("device fully booted after %lums\n", AP_HAL::millis());
+        return true;
+    }
+    return false;
+}
+
+// error handler for OSD simulation
+void AP_RunCam::simulation_OSD_cable_failed(RunCamResponse* response)
+{
+    _waiting_device_response = false;
+    if (response->_command == RCDEVICE_PROTOCOL_COMMAND_5KEY_CONNECTION) {
+        uint8_t operationID = response->_param;
+        if (operationID == RCDEVICE_PROTOCOL_5KEY_CONNECTION_CLOSE) {
+            return;
+        }
+    }
+}
+
+// constructor for a response structure
+AP_RunCam::RunCamResponse::RunCamResponse(AP_RunCam* device, uint8_t commandID, uint8_t param,
+    uint32_t timeout, uint16_t maxRetryTimes, void *userInfo, rcdevice_resp_parse_func_t parserFunc)
+    : _recv_buf(device->_recv_buf),
+    _command(commandID),
+    _max_retry_times(maxRetryTimes),
+    _timeout_ms(timeout),
+    _device(device),
+    _param(param),
+    _user_info(userInfo),
+    _parser_func(parserFunc)
+{
+    _timeout_timestamp = AP_HAL::millis() + timeout;
```

Comment:
Adding times is typically a mistake due to integer overflow issues.

Reply:
Fixed